### DISCUSSION TITLE
Block definition based on phase field evolution

### DISCUSSION THREAD

**Discussion by Bala-1005** [idaholab/moose#16715] [Q&A General] [answered]
Hello everyone,

I am trying to couple phase field and tensor mechanics module. One of the phases is a liquid for which I need the tensor mechanics action to be inactive. I thought of making only the solid phase active for tensor mechanics. 
How can I define blocks in my domain based on the phase-field variable? That is block 0 being the liquid and block 1 being solid.

Thanks,
Bala

**Comment by jiangwen84** [comment, marked as answer]
Currently MOOSE cannot assign/change blocks with a variable value, and might be it is possible in the future, see https://github.com/idaholab/moose/discussions/16635 @aeslaughter  

For your application, I might want to solve the tensor mechanics problem on the whole domain, and use a small stiffness (i.e. Young's modulus) for the fluid phase. Below is an example to compute a `two phase stress`: 
 
 [./global_stress]
    type = TwoPhaseStressMaterial
    base_A = solid
    base_B = liquid
  [../]
  [./switching]
    type = SwitchingFunctionMaterial
    eta = c #  phase field variable for liquid
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '385000 0.23'
    fill_method = symmetric_isotropic_E_nu
    base_name = solid
  [../]
  [./elasticity_tensor_void]
    type = ComputeElasticityTensor
    C_ijkl = '3.85 0.23'
    fill_method = symmetric_isotropic_E_nu
    base_name = liquid
  [../]

**Reply by Bala-1005** [reply]
Thank you @jiangwen84 . I will try to implement this. I did try to run it using the Khachaturyan's scheme (as in modules/combined/examples/phase_field-mechanics/kks_mechanics_KHS.i). Here the elasticity tensors evolved with phases. But I have convergence issues.

Bala

**Comment by aeslaughter** [comment]
Dynamic subdomains based on a variable is something I would like to implement. But, there currently is no plan to perform this work. Perhaps that will change in the future.

**Comment by Bala-1005** [comment]
Using a small stiffness for liquid compared to the solid phase seems to cause a lot of convergence issues. Is there any other way to go about it? @aeslaughter @jiangwen84 

Thanks,
Bala

**Reply by hugary1995** [reply]
I have a dynamic element subdomain modifier under the works. Not sure when I will have time to merge that into moose, probably the next month or so.

There is an existing ActivateElementByCoupled Userobject that you can try.

**Reply by Bala-1005** [reply]
Thank you @hugary1995 . I just tried it and I got the following error.

"A 'ActivateElementsCoupled' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable."

How do I register this object?

Thanks,
Bala
